skyportal/handlers/api/user_obj_list.py

```diff
@@ -0,0 +1,356 @@
+            return self.error(f'User "{user_id}" does not exist!')
+
+        try:
+            user_id = int(user_id)
+        except ValueError:
+            return self.error("Invalid `user_id` parameter; unable to parse to integer")
+
+        # verify that poster has read access to user_id's lists
+        if (
+            self.associated_user_object.id != user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error('Insufficient permissions to access this listing. ')
+
+        list_name = self.get_query_argument("listName", None)
+
+        query = DBSession().query(Listing).filter(Listing.user_id == user_id)
+
+        if list_name is not None:
+            query = query.filter(Listing.list_name == list_name)
+
+        return self.success(data=query.all())
+
+    def add_listing(self, error_if_exists):
+        """
+        Add the listing given in the request body.
+        Two different behaviors when error_if_exists is true/false:
+        - true: will throw an error if listing exists (POST)
+        - false: will ignore if listing already exists (PUT)
+        """
+
+        data = self.get_json()
+
+        schema = Listing.__schema__()
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error(f'Invalid/missing parameters: {e.normalized_messages()}')
+
+        user_id = int(data.get('user_id'))
+
+        if User.query.get(user_id) is None:  # verify that user exists
+            return self.error(f'User "{user_id}" does not exist!')
+
+        # verify that poster has write access to user_id's lists
+        if (
+            not self.associated_user_object.id == user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error('Insufficient permissions to access this listing. ')
+
+        obj_id = data.get('obj_id')
+        if Obj.query.get(obj_id) is None:  # verify that object exists!
+            return self.error(f'Object "{obj_id}" does not exist!')
+
+        list_name = data.get('list_name')
+        if not re.search(r'^\w+', list_name):
+            return self.error(
+                "Input `list_name` must begin with alphanumeric/underscore"
+            )
+
+        query = (
+            DBSession()
+            .query(Listing)
+            .filter(
+                Listing.user_id == user_id,
+                Listing.obj_id == obj_id,
+                Listing.list_name == list_name,
+            )
+        )
+
+        # what to do if listing already exists...
+        if query.first() is not None:
+            if error_if_exists:
+                return self.error(
+                    f'Listing already exists with user_id={user_id}, obj_id={obj_id} and list_name={list_name}'
+                )
+            else:
+                return self.success(data={'listing_id': query.first().id})
```

Comment:
incorrect logic here. on a PUT, we should merge update. I.e., overwrite the existing listing with the new info.

Reply:
this simply returns the old info and discards the new.

Reply:
If user, obj and list name are identical, no need for an update of existing record. Unless I am missing something?